### PR TITLE
MNT: Prevent users from erroneously using legend label API on Axis

### LINKED ISSUE #27971: [Bug]: ax.xaxis.set_label(...) doesn't set the x-axis label
### Bug summary

I'm used to using the Axis API like `ax.xaxis.set_major_formatter(...)` etc., so to be consistent it would be nice if one could do `ax.xaxis.set_label(...)` instead of `ax.set_xlabel(...)`. It seems like `ax.xaxis.set_label(...)` does not label the axis however.

### Code for reproduction

```Python
import matplotlib.pyplot as plt

fig, ax = plt.subplots()
ax.xaxis.set_label("x-label")
```

### Actual outcome

Screenshot 2024-03-24 at 10 02 10

### Expected outcome

Screenshot 2024-03-24 at 10 02 26

### Additional information

_No response_

### Operating system

_No response_

### Matplotlib Version

3.8.3

### Matplotlib Backend

_No response_

### Python version

_No response_

### Jupyter version

_No response_

### Installation

conda

### PR DESCRIPTION
**Pull request by timhoffm**
Closes #27971.

For a complete explanation see https://github.com/matplotlib/matplotlib/issues/27971#issuecomment-2016955731

### CODE REVIEW

**Comment by timhoffm** [comment]
Need to fix internal usage of `Axis.get_label`.

File: lib/matplotlib/axis.py [hunk 1]
```diff
@@ -1421,8 +1421,21 @@ def get_gridlines(self):
...
+
     def get_label(self):
-        """Return the axis label as a Text instance."""
+        """
+        Return the axis label as a Text instance.
+
+        .. admonition:: Discouraged
+
+           This overrides `.Artist.get_label`, which is for legend labels, with a new
+           semantic. It is recommended to use the attribute `.Axis.label` instead.
```
Comment (timhoffm) [inline lib/matplotlib/axis.py:1437]: I've been defensive here and only discouraged but not deprecated this. It's been working for users and I'm not clear whether a deprecation is worth it. We can always add a deprecation later.

File: lib/matplotlib/axis.py [hunk 2]
```diff
@@ -1421,8 +1421,21 @@ def get_gridlines(self):
         return cbook.silent_list('Line2D gridline',
                                  [tick.gridline for tick in ticks])
 
+    def set_label(self, s):
+        """Assigning legend labels is supported. Raises RuntimeError."""
```
Comment (QuLogic) [inline lib/matplotlib/axis.py:1425]: Is supported or is not supported?

File: lib/matplotlib/axis.py [hunk 3]
```diff
@@ -1421,8 +1421,21 @@ def get_gridlines(self):
...
+
     def get_label(self):
-        """Return the axis label as a Text instance."""
+        """
+        Return the axis label as a Text instance.
+
+        .. admonition:: Discouraged
+
+           This overrides `.Artist.get_label`, which is for legend labels, with a new
+           semantic. It is recommended to use the attribute ``Axis.label`` instead.
```
Comment (QuLogic) [inline lib/matplotlib/axis.py:1437]: Can we get `Axis.label` documented so this will link?

Comment (timhoffm) [reply lib/matplotlib/axis.py:1437]: The way attributes are currently documented - Attributes section in the class docstring (https://matplotlib.org/stable/api/axis_api.html#matplotlib.axis.Axis) - they cannot be referenced. I believe we must swich to something like #28825. But before going full scale, we should test it on `Axes` in #28825. This is far beyond this PR. I would like to get this PR into 3.10, but don't want to push the documentation restructuring for it. For now, I suggest that literal is good enough and we can do attribute linking when there is time.

**Review by QuLogic — APPROVED** [review]

**Review by ksunden — APPROVED** [review]

**Comment by QuLogic** [comment]
It looks like we missed that this broke docs?
```
/home/circleci/project/lib/mpl_toolkits/mplot3d/axis3d.py:docstring of matplotlib.artist.Axis.set:24: WARNING: py:meth reference target not found: matplotlib.axis.Axis.set_label [ref.meth]
/home/circleci/project/lib/matplotlib/projections/polar.py:docstring of matplotlib.artist.RadialAxis.set:24: WARNING: py:meth reference target not found: matplotlib.axis.Axis.set_label [ref.meth]
/home/circleci/project/lib/matplotlib/projections/polar.py:docstring of matplotlib.artist.ThetaAxis.set:24: WARNING: py:meth reference target not found: matplotlib.axis.Axis.set_label [ref.meth]
```

**Pull request merged**